vscode-extension/editor/src/VSCodeEditor.tsx

```diff
@@ -0,0 +1,449 @@
+    if (aiConfigServerUrl !== "") {
+      // This is less important for the first load, but when the webview gets dehydrated and rehydrated,
+      // we'll get the aiConfigServerUrl from the webview state. This will trigger a reload of the config.
+      loadConfig();
+    }
+  }, [aiConfigServerUrl, loadConfig]);
+
+  useEffect(() => {
+    loadConfig();
+  }, [loadConfig]);
+
+  // TODO: saqadri - this should be done extension host-side
+  const setupTelemetryIfAllowed = useCallback(async () => {
+    const isDev = (process.env.NODE_ENV ?? "development") === "development";
+
+    // Don't enable telemetry in dev mode because hot reload will spam the logs.
+    if (isDev) {
+      return;
+    }
+
+    const res = await ufetch.get(
+      ROUTE_TABLE.GET_AICONFIGRC(aiConfigServerUrl),
+      {}
+    );
+
+    const enableTelemetry = res.allow_usage_data_sharing;
+
+    if (enableTelemetry) {
+      datadogLogs.init({
+        clientToken: "pub356987caf022337989e492681d1944a8",
+        env: process.env.NODE_ENV ?? "development",
+        service: "aiconfig-editor",
+        site: "us5.datadoghq.com",
+        forwardErrorsToLogs: true,
+        sessionSampleRate: 100,
+      });
+
+      datadogLogs.setGlobalContextProperty("mode", MODE);
+    }
+  }, [aiConfigServerUrl]);
+
+  useEffect(() => {
+    setupTelemetryIfAllowed();
+  }, [setupTelemetryIfAllowed]);
+
+  const getModels = useCallback(
+    async (search: string) => {
+      // For now, rely on caching and handle client-side search filtering
+      // We will use server-side search filtering for Gradio
+      const res = await ufetch.get(ROUTE_TABLE.LIST_MODELS(aiConfigServerUrl));
+      const models = res.data;
+      if (search && search.length > 0) {
+        const lowerCaseSearch = search.toLowerCase();
+        return models.filter(
+          (model: string) =>
+            model.toLocaleLowerCase().indexOf(lowerCaseSearch) >= 0
+        );
+      }
+      return models;
+    },
+    [aiConfigServerUrl]
+  );
+
+  const addPrompt = useCallback(
+    async (promptName: string, promptData: Prompt, index: number) => {
+      const res = await ufetch.post(ROUTE_TABLE.ADD_PROMPT(aiConfigServerUrl), {
+        prompt_name: promptName,
+        prompt_data: promptData,
+        index,
+      });
+
+      if (vscode) {
+        notifyDocumentDirty(vscode);
+      }
+
+      return res;
+    },
+    [aiConfigServerUrl, vscode]
+  );
+
+  const deletePrompt = useCallback(
+    async (promptName: string) => {
+      const res = await ufetch.post(
+        ROUTE_TABLE.DELETE_PROMPT(aiConfigServerUrl),
+        {
+          prompt_name: promptName,
+        }
+      );
+
+      if (vscode) {
+        notifyDocumentDirty(vscode);
+      }
+
+      return res;
+    },
+    [aiConfigServerUrl, vscode]
+  );
+
+  const clearOutputs = useCallback(async () => {
+    const res = await ufetch.post(
+      ROUTE_TABLE.CLEAR_OUTPUTS(aiConfigServerUrl),
+      {}
+    );
+
+    if (vscode) {
```

Comment:
Not needed here, but one nice way to reduce the duplication would be to implement a new hook, `useDocumentActionCallback` which calls the passed-in callback and then does the notifyDocumentDirty(vscode) call

```
export function useDocumentActionCallback(callbackFn, dependencies) {
  const { vscode } = useContext(WebviewContext);
  return useCallback(async (...args) => {
    const res = await callbackFn(...args);
    if (vscode) {
      notifyDocumentDirty(vscode);
    }
    return res;
  }, [dependencies]);
}
```

But no promise this code works, so definitely a follow-up action if at all